src/RBDyn/Jacobian.cpp

```diff
@@ -81,17 +154,33 @@ static inline const Eigen::MatrixXd & jacobian_(const MultiBody & mb,
 
   for(std::size_t index = 0; index < jointsPath.size(); ++index)
   {
-    int i = jointsPath[index];
+    const auto i = static_cast<size_t>(jointsPath[index]);
 
     sva::PTransformd X_i_N = X_0_p * mbc.bodyPosW[i].inv();
 
     for(int dof = 0; dof < joints[i].dof(); ++dof)
     {
-      jac.col(curJ + dof).noalias() = (X_i_N * (sva::MotionVecd(mbc.motionSubspace[i].col(dof)))).vector();
+      if(!reverseJoints[index])
+      {
+        jac.col(curJ + dof).noalias() = (X_i_N * (sva::MotionVecd(mbc.motionSubspace[i].col(dof)))).vector();
+      }
+      else
+      {
+        // The joint motion is seen from child body to parent body, so has inverted effect
+        jac.col(curJ + dof).noalias() = (X_i_N * (-sva::MotionVecd(mbc.motionSubspace[i].col(dof)))).vector();
+      }
     }
 
     curJ += joints[i].dof();
   }
+  // Change Jacobian base : root of the tree --> reference body
+  if(ref_index != -1)
+  {
+    auto dof_count = jac.cols();
+    const auto joint_index = static_cast<size_t>(jointsPath[0]);
+    jac.block(0, 0, 3, dof_count) = mbc.bodyPosW[joint_index].rotation() * jac.block(0, 0, 3, dof_count);
+    jac.block(3, 0, 3, dof_count) = mbc.bodyPosW[joint_index].rotation() * jac.block(3, 0, 3, dof_count);
```

Comment:
I think it should be `ref_index` here and not `joint_index` as this is mixing joint and body indexes.

Also by construction the root of the tree is the reference body so I'm not even sure you need to apply this transformation.

Reply:
If I'm not mistaken, the transformation is needed because you can now choose to use any body of the tree as reference

Reply:
@gergondet you're right I think I mixed up joint and body indexed, so ```ref_index``` should be used instead. As @BenjaminNavarro mentioned, with the new feature the reference body is no longer necessarily the root of the tree but can be any body in the chain.

Reply:
After thinking on it, I think the root of the tree is irrelevant here as the root disappears (only the relative transformation between a joint in the chain and the jacobian's body matters) but I might be missing something (perhaps it will be clearer with tests)